ntoskrnl/io/iomgr/deviface.c

```diff
@@ -304,7 +305,190 @@ IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
+                        Size,
+                        &Size);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("ZwQueryKey() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    /* Enumerate subkeys (i.e. the different device objects) */
+    for (i = 0; i < FullInfo->SubKeys; i++)
+    {
+        Status = ZwEnumerateKey(InterfaceKey,
+                                i,
+                                KeyBasicInformation,
+                                NULL,
+                                0,
+                                &Size);
+        if (Status == STATUS_NO_MORE_ENTRIES)
+        {
+            break;
+        }
+        else if (!NT_SUCCESS(Status) && Status != STATUS_BUFFER_TOO_SMALL)
+        {
+            DPRINT1("ZwEnumerateKey() failed with status 0x%08lx\n", Status);
+            goto cleanup;
+        }
+
+        BasicInfo = ExAllocatePoolWithTag(PagedPool,
+                                          Size,
+                                          TAG_IO);
+        if (!BasicInfo)
+        {
+            DPRINT1("ExAllocatePoolWithTag() failed\n");
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto cleanup;
+        }
+
+        Status = ZwEnumerateKey(InterfaceKey,
+                                i,
+                                KeyBasicInformation,
+                                BasicInfo,
+                                Size,
+                                &Size);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("ZwEnumerateKey() failed with status 0x%08lx\n", Status);
+            goto cleanup;
+        }
+
+        /* Check BasicInfo->Name */
+        RtlInitUnicodeString(&BasicInfoName, NULL);
+        RtlInitUnicodeString(&InLinkName, NULL);
+
+        BasicInfoName.Length = wcslen(BasicInfo->Name);
+        BasicInfoName.MaximumLength = wcslen(BasicInfo->Name);
+        BasicInfoName.Buffer = BasicInfo->Name;
+
+        AliasKeyName.Length = 0;
+        AliasKeyName.MaximumLength += BasicInfoName.Length;
+        AliasKeyName.Buffer = ExAllocatePoolWithTag(NonPagedPool,
+                                                    AliasKeyName.MaximumLength,
+                                                    TAG_IO);
+        if (!AliasKeyName.Buffer)
+        {
+            DPRINT1("ExAllocatePoolWithTag() failed\n");
+            Status = STATUS_INSUFFICIENT_RESOURCES;
+            goto cleanup;
+        }
+
+        Status = RtlAppendUnicodeStringToString(&AliasKeyName, &BasicInfoName);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+            goto cleanup;
+        }
+
+        pdest = wcsstr(SymbolicLinkName->Buffer, L"Services\\");
+
+        InLinkName.Length = wcslen(pdest);
+        InLinkName.MaximumLength = wcslen(pdest);
+        InLinkName.Buffer = pdest;
+
+        if (RtlCompareUnicodeString(&BasicInfoName,
```

Comment:
Does this mean you know the name of the subkey you're looking for? In that case, just open it directly, no need to enumerate anything.